### PR TITLE
Auto-expand replicas when adding or removing nodes

### LINKED ISSUE #1873: Nodes drop their copy of auto-expanded data when coming up, only to sync it again
When you have an index with `index.auto_expand_replicas=0-all` running on 3 nodes and you bring down one node the number of replicas will be reduced by the master from 2 to 1. Then when the node that just went down comes up again ElasticSearch on that node will will:
1. Go up, notice that the number of replicas for that index is 1, and promptly drop its own data as redundant
2. The master will notice that it has a new node in the cluster, set the number of replicas to 2.
3. The node that just dropped its data will now have the data it just dropped re-synced to it.

Instead ElasticSearch should:
1. Go up, wait for the master to adjust the number of replicas if needed
2. Only after that's done drop anything, if needed.
3. Not re-sync any data since it didn't drop the data in the brief interim when the master was adjusting the number of replicas from 1 to 2.

This'll aid recovery time where you have a setup where a relatively small index is available on all the nodes for capacity reasons, and you bring up a new node that should serve search requests right away.

### PR DESCRIPTION
**Pull request by ywelsch**
Auto-expands replicas in the same cluster state update (instead of a follow-up reroute) where nodes are added or removed. 

Fixes #1873

### CODE REVIEW

**Review by bleskes — APPROVED** [review]
LGTM. +1 to move it into the allocation service. I left some nits and questions. Thanks for picking this up!

File: server/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java [hunk 1]
```diff
@@ -80,6 +93,24 @@ int getMaxReplicas(int numDataNodes) {
...
+            final int min = getMinReplicas();
+            final int max = getMaxReplicas(numDataNodes);
+            int numberOfReplicas = numDataNodes - 1;
+            if (numberOfReplicas < min) {
+                numberOfReplicas = min;
+            } else if (numberOfReplicas > max) {
+                numberOfReplicas = max;
+            }
+
+            if (numberOfReplicas >= min && numberOfReplicas <= max) {
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/metadata/AutoExpandReplicas.java]: yeah, if (only) min <= max. Unfortunately we never validated this condition anywhere (I would have expected this during setting validation, but to my surprise, it was not there), and I didn't feel like making a breaking change in this PR by strengthening the validation logic. For now, just as before, if min > max, the auto-expand replicas will be ignored.

**Comment by ywelsch** [comment]
Thanks @bleskes

**Comment by ywelsch** [comment]
@lukas-vlcek The issue resulted from how two distinct components in the system interacted: the auto-expand replica logic, and the shard deletion logic. The way the shard deletion logic works is as follows: When a data node receives a cluster state where a shard is fully allocated (i.e. primary + all replicas are active) and that node has some unallocated data for that shard on disk, it proceeds to delete said data as there is no need to keep this extra copy of the data around. The way the auto-expand replica logic used to work was as follows: This component hooked into the cluster lifecycle, waiting for cluster state changes, comparing the number of nodes in the cluster with the number of auto-expanded replicas and then submitting a settings update to the cluster to adjust the number_of_replicas if the two got out of sync. The way these two components now interacted resulted in the following behavior: Assume a 5 node cluster and an index with auto-expand-replicas set to `0-all`. In that case, `number_of_replicas` is expanded to 4 (i.e. there are 4 replica copies for each primary). Now assume that one node drops out only to later on rejoin the cluster. When the node drops from the cluster, it is removed from the cluster state. In a follow-up settings update initiated by the auto-expand-logic, `number_of_replicas` for the index is adjusted from 4 to 3. When the node rejoins the cluster, it is added to the cluster state by the master and receives this cluster state. The cluster state now has 5 nodes again, but the `number_of_replicas` is not expanded yet from 3 to 4, as this is only triggered shortly after by the auto-expand component (running on the active master). The shard deletion logic on the data node now sees that the shard is fully allocated (`number_of_replicas` set to 3 with all shard copies active). It then proceeds to delete the extra local shard copy, just before it receives the updated cluster state where `number_of_replicas` is now adjusted to 4 again. It therefore has to resync the full data from the primary again as the local copy was completely wiped.

The two main options to fix this were:
- make the shard deletion logic auto-expand replicas aware, accounting for the inconsistencies in the cluster state.
- ensure that cluster states are always consistent w.r.t the auto-expansion of replicas.

For better separation of concerns, and to have stronger consistency guarantees on cluster states, I opted for the second kind of fix. With this PR, the auto-expansion of replicas is done in the same cluster state update where nodes are added or removed from the cluster state. This means that when the newly joining node receives its first cluster state, `number_of_replicas` is already correctly expanded to 4, and the node is aware that the shard is not fully allocated, and therefore does not proceed to delete its local shard copy. The tests in this PR check that the consistency guarantees in the cluster state (`number_of_replicas` is properly expanded at all times) cannot be violated. The shard deletion logic is already separately tested, working on the assumption of a consistent cluster state.

**Pull request merged**